### DISCUSSION TITLE
AWS Lambda Best Practices For Connection Pooling

### DISCUSSION THREAD

**Discussion by bestickley** [porsager/postgres#610] [Show and tell]
Hi, I'd like to document AWS Lambda best practices for connections with this library. Based on the article, [Handling AWS Lambda database connections](https://solidstudio.io/blog/aws-handle-database-connection) I've come up with the following config:
```ts
{
  // https://solidstudio.io/blog/aws-handle-database-connection
  max: 1, // number of connections
  idle_timeout: 1, // seconds
}
```

Know a better configuration? Please share!

**Comment by porsager** [comment]
Would be nice..

I think It'd be best if you could use the certificate they provide https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/PostgreSQL.Concepts.General.SSL.html

**Reply by bestickley** [reply]
Great point, @porsager. Let me specify this to connection pooling guidance.

**Comment by jamesopstad** [comment]
I'm looking for advice on using this library with AWS Lambda and PgBouncer. I currently have PgBouncer set to transaction mode, which seems to be the best option for connecting from lambda.

Is it safest to create the `sql` instance inside the handler function and explicitly call `.end()` or to create the `sql` instance outside the handler function and use a very short `idle_timeout`? With lambda, there is the risk of zombie connections if they are kept alive between invocations because of the freeze/thaw process.

I see from the docs that it is recommended to set `prepare: false` when using PgBouncer in transaction mode. Is it also recommended to set `fetch_types: false`? I'm not clear on the impact of this.

Are there any other settings I need to consider? Sequelize has very detailed documentation on usage with AWS Lambda (https://sequelize.org/docs/v6/other-topics/aws-lambda/) so it might be worth including something similar.

Thanks in advance.

**Reply by bestickley** [reply]
I don't have experience with PgBouncer, but hopefully someone else does and can help!

**Reply by porsager** [reply]
@jamesopstad yes, the issue with lambdas and edge functions is the chatty nature of PostgreSQL startup message (pgbouncer or not). If you are not using array types, I would also recommend setting `fetch_types` to false.

Setting idle_timeout to 0 is fine if you're only doing one request, but if you have more advanced logic, and perhaps also doing other async things inbetween requests, you should go with sql.end().

**Reply by jamesopstad** [reply]
Thanks. It sounds like creating the `sql` instance inside the handler and then calling `sql.end()` before returning the data is the way to go. In that case I guess`idle_timeout` should be a few seconds. I am executing queries in parallel within the handler so I think I'll need to set `max` to something higher than 1 but not so high that concurrent lambdas could open too many connections. Does that sound right?

**Reply by porsager** [reply]
Oh - you can do `const sql = postgres()` at the top level if that looks better for you. Even have it in an imported file as suggested in the docs. Postgres.js is lazy, so no connection happens until you actually make a query 👍

If your queries are parallizable at the db level, then yes, more than one connection might make sense, but I would probably not do that myself at first / only if benchmarks showed it was faster. Remember that the startup chattyness will happen for each connection, so there is a big chance it will be slower if some of the parallel queries execute really fast. Postgres.js has pipelining, so it actually sends all queries to postgres over a single connection even though responses haven't been received yet.

**Reply by jamesopstad** [reply]
> Oh - you can do `const sql = postgres()` at the top level if that looks better for you. Even have it in an imported file as suggested in the docs. Postgres.js is lazy, so no connection happens until you actually make a query 👍

Ah, I see. I had assumed that after calling `sql.end()` the instance could no longer be used.

> If your queries are parallizable at the db level, then yes, more than one connection might make sense, but I would probably not do that myself at first / only if benchmarks showed it was faster. Remember that the startup chattyness will happen for each connection, so there is a big chance it will be slower if some of the parallel queries execute really fast.

I had thought that PgBouncer would help mitigate the cost of initialising connections. I see, however, that if it doesn't already have connections open to the database this could be slower. I'll try benchmarking as you suggest.

> Postgres.js has pipelining, so it actually sends all queries to postgres over a single connection even though responses haven't been received yet.

So, if I have multiple independent queries inside `Promise.all` and `max` is set to `1`, will the queries be executed in series but continue to the next query as soon as each is sent?

Thanks for taking the time to answer my questions. It's very helpful.

**Reply by porsager** [reply]
@jamesopstad sorry for the delay here.

You are correct about the Promise.all scenario. They will be pipelined over that same connection.

**Reply by jamesopstad** [reply]
> > Oh - you can do `const sql = postgres()` at the top level if that looks better for you. Even have it in an imported file as suggested in the docs. Postgres.js is lazy, so no connection happens until you actually make a query 👍
> 
> Ah, I see. I had assumed that after calling `sql.end()` the instance could no longer be used.

By the way, I tried this and it was as I suspected. Declaring `sql` at the top level and then calling `sql.end()` in the handler resulted in a 'CONNECTION_ENDED' error on the next invocation.

**Reply by porsager** [reply]
Well that must mean the same instance is handling several requests, so why end the connection?

**Reply by jamesopstad** [reply]
It comes back to my original question really. I'm not sure if it's safe to keep database connections open between requests. Lambda reuse is non-deterministic and my understanding is that this point I raised could be a concern:

> With lambda, there is the risk of zombie connections if they are kept alive between invocations because of the freeze/thaw process.

I could be wrong though. If connections can survive the freeze/thaw process without issue then keeping the connections open makes sense. There is quite a lot of conflicting advice out there so I'm just reaching out for clarity on the best approach.

**Reply by porsager** [reply]
ah I see.

I think the only sane approach would be doing actual benchmarks. I can't imagine it would make sense to close the connection after each request tbh. A low idle timeout of 1 second sounds reasonable

**Reply by arjunyel** [reply]
@jamesopstad what did you end up deciding?
